enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1705,6 +1706,209 @@ func (z *zstdCompressor) Close() error {
+		return nil, err
+	}
+	if len(missing) > 0 {
+		return nil, status.NotFoundError("Chunks were missing")
+	}
+	pr, pw := io.Pipe()
+	go func() {
+		// read chunks and shove them down the pipe
+		for _, resourceName := range chunkedMD.GetResource() {
+			buf, err := p.Get(ctx, resourceName)
+			if err != nil {
+				pw.CloseWithError(err)
+				return
+			}
+			if _, err := pw.Write(buf); err != nil {
+				pw.CloseWithError(err)
+				return
+			}
+		}
+		pw.Close()
+	}()
+	return pr, nil
+}
+
+type cdcWriter struct {
+	ctx           context.Context
+	pc            *PebbleCache
+	db            pebbleutil.IPebbleDB
+	r             *resource.ResourceName
+	pr            *io.PipeReader
+	pw            *io.PipeWriter
+	done          chan struct{}
+	writtenChunks []*resource.ResourceName
+}
+
+func (p *PebbleCache) newCDCWriter(ctx context.Context, r *resource.ResourceName) (*cdcWriter, error) {
+	db, err := p.leaser.DB()
+	if err != nil {
+		return nil, err
+	}
+	// db.Close is called in p.Close()
+
+	pr, pw := io.Pipe()
+	cdcw := &cdcWriter{
+		ctx:  ctx,
+		pc:   p,
+		db:   db,
+		r:    r,
+		pr:   pr,
+		pw:   pw,
+		done: make(chan struct{}, 0),
+	}
+
+	opts := fastcdc.Options{
+		MinSize:     int(p.maxInlineFileSizeBytes),
+		AverageSize: 1 * 1024 * 1024,
+		MaxSize:     4 * 1024 * 1024,
+	}
+	chunker, err := fastcdc.NewChunker(pr, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			chunk, err := chunker.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				pr.CloseWithError(err)
+				break
+			}
+			start := time.Now()
+			if err := cdcw.writeRawChunk(chunk); err != nil {
+				log.Errorf("Error writing raw chunk: %s", err)
+				pr.CloseWithError(err)
+				break
+			}
+			log.Printf("Took %s to writeRawChunk", time.Since(start))
+		}
+		close(cdcw.done)
+	}()
+	return cdcw, nil
+}
+
+func (cdcw *cdcWriter) writeRawChunk(chunk fastcdc.Chunk) error {
+	ctx := cdcw.ctx
+	p := cdcw.pc
+
+	newBuf := compression.CompressZstd(nil, chunk.Data)
+	d, err := digest.Compute(bytes.NewReader(newBuf))
+	if err != nil {
+		return err
+	}
+	r := &resource.ResourceName{
+		Digest:       d,
+		InstanceName: cdcw.r.GetInstanceName(),
+		Compressor:   repb.Compressor_ZSTD,
+		CacheType:    resource.CacheType_CAS,
+	}
+	fileRecord, err := p.makeFileRecord(ctx, r)
+	if err != nil {
+		return err
+	}
+	key, err := p.fileStorer.PebbleKey(fileRecord)
+	if err != nil {
+		return err
+	}
+	wcm := p.fileStorer.InlineWriter(ctx, d.GetSizeBytes())
+	written, err := wcm.Write(newBuf)
+	if err != nil {
+		return err
+	}
+	if err := wcm.Close(); err != nil {
+		return err
+	}
+
+	now := time.Now().UnixMicro()
+	md := &rfpb.FileMetadata{
+		FileRecord:      fileRecord,
+		StorageMetadata: wcm.Metadata(),
+		StoredSizeBytes: int64(written),
+		LastAccessUsec:  now,
+		LastModifyUsec:  now,
+	}
+	if err := p.writeMetadata(ctx, cdcw.db, key, md); err != nil {
+		return err
+	}
+	cdcw.writtenChunks = append(cdcw.writtenChunks, r)
```

Comment:
I wonder if we could parallelize the lower part of  `writeRawChunk()` in an ErrGroup by moving this `writtenChunks` accumulation up the stack, inside `newCDCWriter()`. But it's probably not worth it for now 🤔 

It definitely took me a bit longer to find this line after the initial read as my intuition was telling me that this should be somewhere in the for-loop above.